robotiq_s_model_control/src/robotiq_s_model_control/s_model_can_client.cpp

```diff
@@ -0,0 +1,421 @@
+        serial_request.data[4] = output.rPRS;
+        if(prevCmd_.find(serial_request.data[3]) == prevCmd_.end() || prevCmd_[serial_request.data[3]]!= serial_request.data[4])
+        {
+            driver_->send(serial_request);
+            prevCmd_[serial_request.data[3]] = serial_request.data[4];
+        }
+
+        serial_request.data[3] = subindex++; // Finger S speed
+        serial_request.data[4] = 0x00;
+        serial_request.data[4] = output.rSPS;
+        if(prevCmd_.find(serial_request.data[3]) == prevCmd_.end() || prevCmd_[serial_request.data[3]]!= serial_request.data[4])
+        {
+            driver_->send(serial_request);
+            prevCmd_[serial_request.data[3]] = serial_request.data[4];
+        }
+
+        serial_request.data[3] = subindex++; // Finger S force
+        serial_request.data[4] = 0x00;
+        serial_request.data[4] = output.rFRS;
+        if(prevCmd_.find(serial_request.data[3]) == prevCmd_.end() || prevCmd_[serial_request.data[3]]!= serial_request.data[4])
+        {
+            driver_->send(serial_request);
+            prevCmd_[serial_request.data[3]] = serial_request.data[4];
+        }
+
+        output_ = output;
+
+    }
+}
+
+SModelCanClient::GripperInput SModelCanClient::readInputs() const
+{
+    if(driver_->getState().isReady())
+    {
+        can::Frame serial_request;
+        serial_request.id = can_id_+0x600;
+        serial_request.dlc = (unsigned char)8;
+        serial_request.is_error = 0;
+        serial_request.is_rtr = 0;
+        serial_request.is_extended = 0;
+        serial_request.data.assign((unsigned char)0);
+
+        serial_request.data[0] = 0x4F;
+        serial_request.data[1] = 0x00;
+        serial_request.data[2] = 0x20;
+
+        for(unsigned int i = 1; i < 16; ++i)
+        {
+            serial_request.data[3] = i;
+            if(!read_mutex.timed_lock(boost::posix_time::time_duration(0,0,1,0)))
+            {
+                std::stringstream err;
+                err<<"unable to read response on index "<<i;
+                read_mutex.unlock();
+                throw std::runtime_error(err.str().c_str());
+            }
+            driver_->send(serial_request);
+        }
+        //! try to lock one more time to wait for last read
+        if(!read_mutex.timed_lock(boost::posix_time::time_duration(0,0,1,0)))
+        {
+            std::stringstream err;
+            err<<"unable to read response on last index";
+            read_mutex.unlock();
+            throw std::runtime_error(err.str().c_str());
+        }
+
+        read_mutex.unlock();
+    }
+    return input_;
+}
+
+SModelCanClient::GripperOutput SModelCanClient::readOutputs() const
+{
+    return output_;
+}
+
+void SModelCanClient::stateCallback(const can::State &s)
+{
+    std::string err;
+    driver_->translateError(s.internal_error, err);
+    if(!s.internal_error)
+    {
+        std::cout<<"State: " <<err <<", asio: "<<s.error_code.message()<<std::endl;
```

Comment:
Please prefer the use of `ROS_INFO`, `ROS_DEBUG`, etc for these components. It's helpful for debugging and plays nice with the logging system.